### PR TITLE
[14.0] Move currencies from shopinvader_locomotive to shopinvader, add endpoints for backend metadata

### PR DESCRIPTION
**Pull request by kevinkhao**
Add routes for:
* currencies
* languages
* titles
* industries
* countries
* combination of all of the above

### CODE REVIEW

File: shopinvader/services/settings.py
Comment (simahawk) [inline shopinvader/services/settings.py]:
```suggestion
        [(["/countries"], "GET")],
```

File: shopinvader/services/settings.py
Comment (simahawk) [inline shopinvader/services/settings.py]:
```suggestion
        [(["/titles], "GET")],
```

File: shopinvader/services/settings.py
Comment (simahawk) [inline shopinvader/services/settings.py]:
```suggestion
    def countries(self):
```
aaand repeat for all endpoint IMO :wink:

File: shopinvader_locomotive/models/shopinvader_backend.py
Comment (simahawk) [inline shopinvader_locomotive/models/shopinvader_backend.py:28]: thanks for moving this! Was in my low prio TODO list since ages :smile:

File: shopinvader/services/settings.py
Comment (simahawk) [inline shopinvader/services/settings.py]:
```suggestion
    _name = "shopinvader.settings.service"
```
no export here

File: shopinvader/services/settings.py
Comment (simahawk) [inline shopinvader/services/settings.py]:
```suggestion
    """Shopinvader service to expose allowed settings"""
```

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by kevinkhao** [comment]
@simahawk done !

**Review by simahawk — APPROVED** [review]

File: shopinvader/services/settings.py
Comment (lmignon) [inline shopinvader/services/settings.py]: @kevinkhao Can you add the response schema for all the methods...
```suggestion

   @property
   def _get_title_schema(self):
       return {
           "name": {"type": "string", "required": True, "nullable": False},
           "id": {"type": "integer", "required": True, "nullable": False}
       }
       
    @restapi.method(
        [(["/titles"], "GET")], 
        output_param=restapi.CerberusValidator("_get_title_schema", is_list=True)
    )
    def titles(self):
        return self._get_titles()
```

File: shopinvader/services/settings.py
Comment (lmignon) [inline shopinvader/services/settings.py]: All these methods should be public.
```suggestion
    @restapi.method(
        [(["/titles"], "GET")],
        auth="public"
    )
```

**Review by lmignon — CHANGES_REQUESTED** [review]

File: shopinvader/services/settings.py
Comment (lmignon) [inline shopinvader/services/settings.py]: @kevinkhao Can you add the explicit decorator PLZ. 
```suggestion
    @restapi.method(
        [(["/"], "GET")],
        output_param=restapi.CerberusValidator(...)
        auth="public",
    )
    def get(self):
```

File: shopinvader/services/settings.py
Comment (lmignon) [inline shopinvader/services/settings.py]: should be public.... Moreover, auth method `api_key` is no more provided by shopinvader. This authentication method is now provided by shopinvader_auth_api_key. This addon declare the REST controller used to serve all the services under the `shopinvader`root path with `api_key` as default auth. https://github.com/shopinvader/odoo-shopinvader/blob/14.0/shopinvader_auth_api_key/controllers/main.py The default auth is applied to all the service methods that doesn't explicitely declare the required authentication method. If you omit the `auth` param from `@restapi.method`, the default auth method defined on the controller registering the services at a given path will be applied. It's a key feature that allows you to serve the same services under 2 root paths with different authentication method. ex: with `shopinvader_auth_jwt` https://github.com/shopinvader/odoo-shopinvader/pull/1031/files#diff-65350ca569ae9b77446b7b6dacfd348e127ad4da850f96b4091a6434f4eaf8afR7

File: shopinvader/services/settings.py
Comment (kevinkhao) [reply shopinvader/services/settings.py:64]: right now I always return a dict that has lists for its values, for example for /titles it will be {'titles': [t1, t2...]} . Unless convention is specifically to return it in list format ?

**Comment by kevinkhao** [comment]
- [x] explicit decorator for get
- [x] add public auth for all
- [x] add response schemas

File: shopinvader/services/settings.py
Comment (kevinkhao) [reply shopinvader/services/settings.py]: I see, thank you for the explanation

**Comment by kevinkhao** [comment]
@lmignon done

File: shopinvader/services/settings.py
Comment (kevinkhao) [reply shopinvader/services/settings.py]: done https://github.com/shopinvader/odoo-shopinvader/pull/1049/commits/7d9cb722cf4d80d0fc2f4f62d82529de76c598cf

Comment (kevinkhao) [reply shopinvader/services/settings.py]: pls help check as I did not use is_list it seems not implemented

File: shopinvader/tests/test_settings.py
Comment (lmignon) [inline shopinvader/tests/test_settings.py]:
```suggestion
class SettingsTest(CommonSettingsCase):
```

Comment (lmignon) [reply shopinvader/tests/test_settings.py]: @kevinkhao Why do we need  AddressTestCase and and AdressCase. There is no value for AddressTestCase

File: shopinvader/tests/test_settings.py
Comment (lmignon) [inline shopinvader/tests/test_settings.py]: should be removes

File: shopinvader/tests/test_settings.py
Comment (lmignon) [inline shopinvader/tests/test_settings.py]: cls.partner is never used. Moreover, do we need a CommonSettingsCase. IMO we don't need it.

**Review by lmignon — CHANGES_REQUESTED** [review]

File: shopinvader/tests/test_settings.py
Comment (kevinkhao) [reply shopinvader/tests/test_settings.py]: removed

File: shopinvader/services/settings.py
Comment (lmignon) [inline shopinvader/services/settings.py]:
```suggestion
        [(["/", "/all"], "GET")],
```

**Review by lmignon — APPROVED** [review]
Thank you @kevinkhao for all this. LGTM once https://github.com/shopinvader/odoo-shopinvader/pull/1049/files#r681864391 is fixed :exclamation:

**Comment by lmignon** [comment]
@kevinkhao tests fails due to https://github.com/odoo/odoo/pull/73831 and fixed by https://github.com/odoo/odoo/pull/74618

**Comment by sbidoul** [comment]
The fix is merged in Odoo, but we need to wait until tomorrow for the oca-ci image to be rebuilt during the night.

**Comment by rousseldenis** [comment]
Tests are still red since elasticsearch lib has been updated.

I did a PR on connector_elasticsearch to limit version: https://github.com/OCA/search-engine/pull/97

**Comment by kevinkhao** [comment]
up

**Pull request merged**
